designs/2023-sharding/README.md

```diff
@@ -0,0 +1,96 @@
+
+## Drawbacks
+
+Introducing sharding in the ESLint package has several potential drawbacks to consider:
+
+- **Increased Complexity**: Sharding adds complexity to the linting process, requiring additional logic for workload division, parallel execution, and result aggregation. This complexity may make the codebase harder to maintain and debug.
+
+- **Resource Consumption**: Sharding involves launching multiple processes/threads to handle each shard concurrently. This can result in increased resource consumption, such as CPU and memory usage, particularly when dealing with a large number of shards or codebases.
+
+- **Potential Overhead**: The overhead of shard coordination and result aggregation may impact the overall performance gain achieved through parallel execution. Careful optimization and benchmarking should be performed to ensure that the benefits of sharding outweigh the associated overhead.
+
+## Backwards Compatibility Analysis
+
+
+The introduction of sharding in the ESLint package should not affect existing ESLint users by default. Sharding should be an opt-in feature, enabled through configuration options or flags. Users who do not wish to utilize sharding can continue to use ESLint as before without any disruption.
+
+## Alternatives
+
+Alternative approaches to improve linting performance in ESLint include:
+
+Manually splitting the project to run in parallel example by @bmish in https://github.com/eslint/eslint/issues/16726#issuecomment-1368059618.
+
+```json
+{
+  "scripts": {
+    "lint:js": "npm-run-all --aggregate-output --continue-on-error --parallel \"lint:js:*\"",
+    "lint:js:dir1": "eslint --cache dir1/",
+    "lint:js:dir2": "eslint --cache dir2/",
+    "lint:js:dir3": "eslint --cache dir3/",
+    "lint:js:other": "eslint --cache --ignore-pattern dir1/ --ignore-pattern dir2/ --ignore-pattern dir3/ .",
+  }
+}
+```
+
+
+## Open Questions
+
+- What process/thread management strategy should be employed to handle the shards effectively?
+    
+- How to allow custom division statergy?
+
+## Help Needed
+
+As a first-time contributor, I would greatly appreciate guidance and assistance from the ESLint core team and the community in implementing the sharding feature. While I have experience as a programmer, contributing to a large-scale project like ESLint involves understanding the codebase, adhering to coding standards, and following the established contribution process.
+
+
+## Frequently Asked Questions
+
+Q: Will enabling sharding affect the linting results or introduce any inconsistencies?
+
+A: Enabling sharding should not affect the linting results or introduce any inconsistencies. Sharding is designed to distribute the workload and execute the linting process independently on each shard. The aggregated results should provide a unified view of the linting issues across all shards, ensuring consistent analysis.
```

Comment:
I think you're right that this won't affect the current analysis. That said, it can limit future kinds of analysis like multi-file linting ([which I've written about here](https://jfmengels.net/multi-file-analysis/), specifically [this section](https://jfmengels.net/multi-file-analysis/#need-to-analyze-everything)), where sharding could make this impossible or at best become a hindrance that would require new solutions (like opting out some rules out of the sharding strategy).
You could end up with different results when turning on/off sharding, which would be pretty bad.
That said, I don't remember whether ESLint aims to implement this one day, but it's like eslint-plugin-import does in a round-about way.